File: java/cours/src/session2/td/heritageMultiple/AlgebreNatParInt.java
package session2.td.heritageMultiple;

import session1.td.Nat;

public interface AlgebreNatParInt extends Nat {

	@Override
	public default Nat zero() {
		return this.creerNatAvecValeur(0);
	}

	@Override
	public default Nat somme(Nat x) {
		return this.creerNatAvecValeur(this.val() + x.val());
	}

	@Override
	public default Nat un() {
		return this.creerNatAvecValeur(1);
	}

	@Override
	public default Nat produit(Nat x) {
		return this.creerNatAvecValeur(this.val() * x.val());
	}


	@Override
	public default Nat modulo(Nat x) {
		return this.creerNatAvecValeur(this.val()%x.val());
	}

	@Override
	public default Nat div(Nat x) {
		return this.creerNatAvecValeur(this.val()/x.val());
	}
	
	public default String representer() {
		return Integer.toString(this.val());
	}
	
	public default boolean estEgal(Object obj) {
		if(!(obj instanceof Nat))
			return false;
		Nat x = (Nat)obj;
		return this.val() == x.val();
	}
}

File: typescript/session2/heritageMultiple/AlgebreNatParInt.ts
import {Nat} from "../../session1/td/Nat/Nat";

export abstract class AlgebreNatParInt extends Nat {

    zero() : Nat {
        return this.creerNatValue(0);
    }

    somme(x : Nat) : Nat {
        return this.creerNatValue(this.getVal() + x.getVal());
    }

    un() : Nat {
        return this.creerNatValue(1);
    }

    produit(x : Nat) : Nat{
        return this.creerNatValue(this.getVal() * x.getVal());
    }


    modulo(x : Nat) : Nat {
        return this.creerNatValue(this.getVal()%x.getVal());
    }

    div(x : Nat) : Nat {
        return this.creerNatValue(this.getVal()/x.getVal());
    }

    representer() : string {
        return '' + this.getVal();
    }

    estEgal(obj : Object) : boolean{
        if(!(obj instanceof Nat))
            return false;
        let x = <Nat> obj;
        return this.getVal() == x.getVal();
    }
}
